ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixParticipant.java

```diff
@@ -454,62 +458,97 @@ boolean setStoppedReplicas(List<String> stoppedReplicas) {
+      PartitionStateChangeListener storageManagerListener =
+          partitionStateChangeListeners.get(StateModelListenerType.StorageManagerListener);
+      if (storageManagerListener != null) {
+        storageManagerListener.onPartitionBecomeBootstrapFromOffline(partitionName);
+      }
+      // 2. take actions in replication manager (add new replica if necessary)
+      PartitionStateChangeListener replicationManagerListener =
+          partitionStateChangeListeners.get(StateModelListenerType.ReplicationManagerListener);
+      if (replicationManagerListener != null) {
+        replicationManagerListener.onPartitionBecomeBootstrapFromOffline(partitionName);
+      }
+      // 3. take actions in stats manager (add new replica if necessary)
+      PartitionStateChangeListener statsManagerListener =
+          partitionStateChangeListeners.get(StateModelListenerType.StatsManagerListener);
+      if (statsManagerListener != null) {
+        statsManagerListener.onPartitionBecomeBootstrapFromOffline(partitionName);
+      }
+    } catch (Exception e) {
+      participantMetrics.offlineCount.addAndGet(offlineCountChange);
+      participantMetrics.errorStateCount.addAndGet(1);
+      throw e;
+    }
+    participantMetrics.offlineCount.addAndGet(offlineCountChange);
+    participantMetrics.bootstrapCount.addAndGet(1);
+    // Here we directly add the partition into set even though it may already exit because the op should be idempotent)
+    localPartitions.add(partitionName);
   }
 
   @Override
   public void onPartitionBecomeStandbyFromBootstrap(String partitionName) {
     PartitionStateChangeListener replicationManagerListener =
         partitionStateChangeListeners.get(StateModelListenerType.ReplicationManagerListener);
     if (replicationManagerListener != null) {
-      replicationManagerListener.onPartitionBecomeStandbyFromBootstrap(partitionName);
-      // after bootstrap is initiated in ReplicationManager, transition is blocked here and wait until local replica has
-      // caught up with enough peer replicas.
       try {
+        replicationManagerListener.onPartitionBecomeStandbyFromBootstrap(partitionName);
+        // after bootstrap is initiated in ReplicationManager, transition is blocked here and wait until local replica has
+        // caught up with enough peer replicas.
         replicaSyncUpManager.waitBootstrapCompleted(partitionName);
       } catch (InterruptedException e) {
         logger.error("Bootstrap was interrupted on partition {}", partitionName);
+        participantMetrics.bootstrapCount.addAndGet(-1);
+        participantMetrics.errorStateCount.addAndGet(1);
         throw new StateTransitionException("Bootstrap failed or was interrupted", BootstrapFailure);
       } catch (StateTransitionException e) {
         logger.error("Bootstrap didn't complete on partition {}", partitionName, e);
+        participantMetrics.bootstrapCount.addAndGet(-1);
```

Comment:
For all the metrics related changes in this file, it might be cleaner to update the participantMetrics in finally block.